syncer/streamer_controller.go

```diff
@@ -0,0 +1,412 @@
+	c.Lock()
+	defer c.Unlock()
+
+	tctx.L().Info("reset global streamer", zap.Stringer("position", pos))
+	return c.resetReplicationSyncer(tctx, pos)
+}
+
+// GetEvent returns binlog event, should only have one thread call this function.
+func (c *StreamerController) GetEvent(tctx *tcontext.Context) (event *replication.BinlogEvent, err error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	ctx, cancel := context.WithTimeout(tctx.Context(), eventTimeout)
+	failpoint.Inject("SyncerEventTimeout", func(val failpoint.Value) {
+		if seconds, ok := val.(int); ok {
+			cancel()
+			ctx, cancel = context.WithTimeout(tctx.Context(), time.Duration(seconds)*time.Second)
+			tctx.L().Info("set fetch binlog event timeout", zap.String("failpoint", "SyncerEventTimeout"), zap.Int("value", seconds))
+		}
+	})
+
+	event, err = c.streamer.GetEvent(ctx)
+	cancel()
+	if err != nil {
+		if err != context.Canceled && err != context.DeadlineExceeded {
+			c.meetError = true
+		}
+
+		return nil, err
+	}
+
+	switch ev := event.Event.(type) {
+	case *replication.RotateEvent:
+		// if is local binlog, binlog's name contain uuid information, need to save it
+		// if is remote binlog, need to add uuid information in binlog's name
+		if !c.setUUIDIfExists(string(ev.NextLogName)) {
+			if len(c.uuidSuffix) != 0 {
+				filename, err := binlog.ParseFilename(string(ev.NextLogName))
+				if err != nil {
+					return nil, terror.Annotate(err, "fail to parse binlog file name from rotate event")
+				}
+				ev.NextLogName = []byte(binlog.ConstructFilenameWithUUIDSuffix(filename, c.uuidSuffix))
+				event.Event = ev
+			}
+		}
+	default:
+	}
+
+	return event, nil
+}
+
+// ReopenWithRetry reopens streamer with retry
+func (c *StreamerController) ReopenWithRetry(tctx *tcontext.Context, pos mysql.Position) error {
+	c.Lock()
+	defer c.Unlock()
+
+	var err error
+	for i := 0; i < maxRetryCount; i++ {
+		err = c.resetReplicationSyncer(tctx, pos)
+		if err == nil {
+			return nil
+		}
+		if needRetryReplicate(err) {
+			tctx.L().Info("fail to retry open binlog streamer", log.ShortError(err))
+			time.Sleep(retryTimeout)
+			continue
+		}
+		break
+	}
+	return err
+}
+
+func (c *StreamerController) closeBinlogSyncer(logtctx *tcontext.Context, binlogSyncer *replication.BinlogSyncer) error {
+	if binlogSyncer == nil {
+		return nil
+	}
+
+	lastSlaveConnectionID := binlogSyncer.LastConnectionID()
+	defer binlogSyncer.Close()
+	if lastSlaveConnectionID > 0 {
+		err := c.fromDB.killConn(lastSlaveConnectionID)
+		if err != nil {
+			logtctx.L().Error("fail to kill last connection", zap.Uint32("connection ID", lastSlaveConnectionID), log.ShortError(err))
+			if !utils.IsNoSuchThreadError(err) {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Restart restarts streamer controller
+func (c *StreamerController) Restart(tctx *tcontext.Context, pos mysql.Position) error {
+	c.Lock()
+	defer c.Unlock()
+
+	c.meetError = false
+	c.closed = false
```

Comment:
Do we need to reset `currentBinlogType` here (or even other fields)?

Reply:
yes, I fix in f05ac87